### PR TITLE
feat: Add imageID status field

### LINKED ISSUE #4372: Add `karpenter.k8s.aws/instance-ami-id` label
### Description

**Observed Behavior**:
Noticed nodes created by Karpenter after we upgraded to 0.29.2 are missing the karpenter.k8s.aws/instance-ami-id label. We have some operation scripts (if a specific AMI has issues for example) that rely on this label so it would be great if we could get it back.

**Expected Behavior**:
Current Karpenter labels added by one of our provisioners.
```
    karpenter.k8s.aws/instance-category: c
    karpenter.k8s.aws/instance-cpu: "16"
    karpenter.k8s.aws/instance-encryption-in-transit-supported: "true"
    karpenter.k8s.aws/instance-family: c5n
    karpenter.k8s.aws/instance-generation: "5"
    karpenter.k8s.aws/instance-hypervisor: nitro
    karpenter.k8s.aws/instance-memory: "43008"
    karpenter.k8s.aws/instance-network-bandwidth: "15000"
    karpenter.k8s.aws/instance-pods: "234"
    karpenter.k8s.aws/instance-size: 4xlarge
    karpenter.sh/capacity-type: on-demand
    karpenter.sh/initialized: "true"
    karpenter.sh/provisioner-name: key-server
    karpenter.sh/registered: "true"
```

**Reproduction Steps** (Please include YAML):
A node created by the same provisioner last week:
```
    karpenter.k8s.aws/instance-ami-id: ami-0158858fcc1c5f6a3
    karpenter.k8s.aws/instance-category: c
    karpenter.k8s.aws/instance-cpu: "16"
    karpenter.k8s.aws/instance-encryption-in-transit-supported: "true"
    karpenter.k8s.aws/instance-family: c5n
    karpenter.k8s.aws/instance-generation: "5"
    karpenter.k8s.aws/instance-hypervisor: nitro
    karpenter.k8s.aws/instance-memory: "43008"
    karpenter.k8s.aws/instance-network-bandwidth: "15000"
    karpenter.k8s.aws/instance-pods: "234"
    karpenter.k8s.aws/instance-size: 4xlarge
    karpenter.sh/capacity-type: on-demand
    karpenter.sh/initialized: "true"
    karpenter.sh/provisioner-name: key-server
    karpenter.sh/registered: "true"
```

**Versions**:
- Chart Version: 0.29.2
- Kubernetes Version (`kubectl version`): 1.23
```
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.17-eks-a5565ad", GitCommit:"23f44c7dd4e8e23856387345e7cb80bf0f93ced6", GitTreeState:"clean", BuildDate:"2023-06-15T21:19:39Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
```

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by gfcroft**
feat: Add karpenter.k8s.aws/instance-ami-id label
Fixes #4372 

**Description**
Add the karpenter.k8s.aws/instance-ami-id label to nodes via retrieval of ami-id value from feet override reponse
- Because a node template may have an existing launch template name set, we have to look up the data for the launch template via an AWS Api call to ec2:DescribeLaunchTemplateVersions
    - Note that there will be extra expense therefore in order to perform this op, so I have added use of the cache for this data
    - Launch template versions seems to be the only viable api to get at the full template data
 - To use this API, we need to add permission to describe launch template versions on the Karpenter controller role; I have update the docs to use ec2:DescribeLaunchTemplate* in the permissions set, instead of ec2:DescribeLaunchTemplates, in order to cover this additional permission (and any similar permissions required in future)
- the only constant I could see we currently have to cover karpenter.k8s.aws/instance-ami-id label is  Have to add in the image id - seems it’s just defined in v1alpha1 ? Hopefully that’s not totally deprecated… 
- By always using imageID as an override parameter, we are dependant now on create fleet request being for a fleet of type ‘instant’ -  see: https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#FleetLaunchTemplateOverridesRequest     
      - We seem to only create ‘instant’ fleets, so this will work for now - https://github.com/aws/karpenter/blob/2dfebc9602a9bf6848616fb3740c0f25a455d650/pkg/providers/instance/instance.go#L209-L209 
   
**How was this change tested?**
requirement for label added to unit tests, also verified by running my against my own eks cluster and seeing that the node labels are indeed added to provisioned nodes appropriately (either via nodes provisioned by karpenter auto-generated launch templates, or via a pre-defined launch template set in aws node template)

**Does this change impact docs?**
- [X] Yes, PR includes docs updates 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: website/content/en/docs/getting-started/getting-started-with-karpenter/cloudformation.yaml [hunk 1]
```diff
@@ -169,7 +169,7 @@ Resources:
                 "ec2:DescribeInstances",
                 "ec2:DescribeInstanceTypeOfferings",
                 "ec2:DescribeInstanceTypes",
-                "ec2:DescribeLaunchTemplates",
+                "ec2:DescribeLaunchTemplate*",
```
Comment (FernandoMiguel) [inline website/content/en/docs/getting-started/getting-started-with-karpenter/cloudformation.yaml:172]: why this change?

Comment (gfcroft) [reply website/content/en/docs/getting-started/getting-started-with-karpenter/cloudformation.yaml:172]: Because a node template may have an existing launch template name set, we have to look up the data for the launch template via an AWS Api call to ec2:DescribeLaunchTemplateVersions

Launch template versions seems to be the only viable api to get at the full template data

To use this API, we need to add permission to describe launch template versions on the Karpenter controller role; I have update the docs to use ec2:DescribeLaunchTemplate* in the permissions set, instead of ec2:DescribeLaunchTemplates, in order to cover this additional permission (and any similar permissions required in future)

Comment (gfcroft) [reply website/content/en/docs/getting-started/getting-started-with-karpenter/cloudformation.yaml:172]: @jonathan-innis oh sorry I wasn't aware of that proposal - sure lets not populate the label in the case where someone is using a custom launch template and sidestep this issue.

File: pkg/providers/instancetype/zz_generated.bandwidth.go [hunk 2]
```diff
@@ -42,6 +42,7 @@ var (
 		"t4g.medium":        256,
 		"m7a.medium":        390,
 		"m7i-flex.large":    390,
+		"r7a.medium":        390,
```
Comment (jonathan-innis) [inline pkg/providers/instancetype/zz_generated.bandwidth.go:45]: I think we can drop updates to the `codegen` that are unrelated to the PR itself.

File: pkg/providers/launchtemplate/launchtemplate.go [hunk 3]
```diff
@@ -92,14 +98,16 @@ func NewProvider(ctx context.Context, cache *cache.Cache, ec2api ec2iface.EC2API
...
 func (p *Provider) EnsureAll(ctx context.Context, nodeClass *v1beta1.NodeClass, nodeClaim *corev1beta1.NodeClaim,
-	instanceTypes []*cloudprovider.InstanceType, additionalLabels map[string]string, tags map[string]string) (map[string][]*cloudprovider.InstanceType, error) {
+	instanceTypes []*cloudprovider.InstanceType, additionalLabels map[string]string, tags map[string]string) ([]*LaunchTemplate, error) {
 
 	p.Lock()
 	defer p.Unlock()
 	// If Launch Template is directly specified then just use it
 	if nodeClass.Spec.LaunchTemplateName != nil {
-		return map[string][]*cloudprovider.InstanceType{ptr.StringValue(nodeClass.Spec.LaunchTemplateName): instanceTypes}, nil
+		templateName := ptr.StringValue(nodeClass.Spec.LaunchTemplateName)
```
Comment (jonathan-innis) [inline pkg/providers/launchtemplate/launchtemplate.go:107]: nit: collapse this into a single line?

File: pkg/providers/launchtemplate/launchtemplate.go [hunk 4]
```diff
@@ -108,14 +116,19 @@ func (p *Provider) EnsureAll(ctx context.Context, nodeClass *v1beta1.NodeClass,
...
-	launchTemplates := map[string][]*cloudprovider.InstanceType{}
+	launchTemplates := []*LaunchTemplate{}
 	for _, resolvedLaunchTemplate := range resolvedLaunchTemplates {
 		// Ensure the launch template exists, or create it
 		ec2LaunchTemplate, err := p.ensureLaunchTemplate(ctx, resolvedLaunchTemplate)
 		if err != nil {
 			return nil, err
 		}
-		launchTemplates[*ec2LaunchTemplate.LaunchTemplateName] = resolvedLaunchTemplate.InstanceTypes
+		launchTemplate := &LaunchTemplate{
```
Comment (jonathan-innis) [inline pkg/providers/launchtemplate/launchtemplate.go:126]: nit: Collapse this and the `append` statement into a single line?

File: pkg/providers/pricing/zz_generated.pricing.go [hunk 5]
```diff
@@ -18,9 +18,9 @@ package pricing
 
 import "time"
 
-// generated at 2023-08-21T14:24:21Z for us-east-1
+// generated at 2023-09-19T22:00:25Z for us-east-1
 
-var initialPriceUpdate, _ = time.Parse(time.RFC3339, "2023-08-21T14:24:21Z")
+var initialPriceUpdate, _ = time.Parse(time.RFC3339, "2023-09-19T22:00:25Z")
```
Comment (jonathan-innis) [inline pkg/providers/pricing/zz_generated.pricing.go:23]: Same comment as the bandwidth gen. I think we can drop the auto-gened changes as part of this PR.

**Comment by jonathan-innis** [comment]
I think you may need a rebase here as well

File: pkg/cloudprovider/cloudprovider.go [hunk 6]
```diff
@@ -358,6 +358,7 @@ func (c *CloudProvider) instanceToNodeClaim(i *instance.Instance, instanceType *
 	}
 	labels[v1.LabelTopologyZone] = i.Zone
 	labels[corev1beta1.CapacityTypeLabelKey] = i.CapacityType
+	labels[v1alpha1.LabelInstanceAMIID] = i.ImageID
```
Comment (jonathan-innis) [inline pkg/cloudprovider/cloudprovider.go:361]:
```suggestion
	labels[v1beta1.LabelInstanceAMIID] = i.ImageID
```

We should add this to the `beta` APIs package since this is what we are taking forward

File: pkg/providers/launchtemplate/launchtemplate.go [hunk 7]
```diff
@@ -113,14 +120,14 @@ func (p *Provider) EnsureAll(ctx context.Context, nodeClass *v1beta1.EC2NodeClas
 	if err != nil {
 		return nil, err
 	}
-	launchTemplates := map[string][]*cloudprovider.InstanceType{}
+	launchTemplates := []*LaunchTemplate{}
```
Comment (jonathan-innis) [inline pkg/providers/launchtemplate/launchtemplate.go:123]:
```suggestion
	var launchTemplates []*LaunchTemplate{}
```

Prefer this syntax: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

**Review by jonathan-innis — DISMISSED** [review]
/karpenter snapshot

**Comment by jonathan-innis** [comment]
@gfcroft Looks like a bunch of the E2E tests failed. Seeing a bunch of this in the logs: `The request failed because the payload for the body of the request is too large`

**Comment by jonathan-innis** [comment]
@gfcroft Responded to the original issue. How do you feel about putting the AMI detail into the NodeClaim status rather than  into the labels? We can keep the labels specific to scheduling and do informational/observability things like this into the status.

**Comment by gfcroft** [comment]
as this change depends on the status field of node claim supporting ImageID which requires an update to karpenter-core (to be accomplished by: https://github.com/aws/karpenter-core/pull/559) I expect tests to fail until that is merged and go.mod is updated accordingly

File: pkg/apis/v1beta1/labels.go [hunk 8]
```diff
@@ -106,6 +106,7 @@ var (
 
 	LabelNodeClass = Group + "/nodeclass"
 
+	LabelInstanceAMIID                        = Group + "/instance-ami-id"
```
Comment (jonathan-innis) [inline pkg/apis/v1beta1/labels.go:109]: Can we drop this label now that we have the status field?

File: pkg/providers/instancetype/zz_generated.bandwidth.go [hunk 9]
```diff
@@ -42,7 +42,6 @@ var (
 		"t4g.medium":        256,
```
Comment (jonathan-innis) [inline pkg/providers/instancetype/zz_generated.bandwidth.go:42]: I think we should drop this filechange for this PR

File: pkg/providers/pricing/zz_generated.pricing.go [hunk 10]
```diff
@@ -18,9 +18,9 @@ package pricing
 
 import "time"
 
-// generated at 2023-09-18T13:06:44Z for us-east-1
```
Comment (jonathan-innis) [inline pkg/providers/pricing/zz_generated.pricing.go:21]: Drop the filechange for this PR

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

File: pkg/providers/instance/instance.go [hunk 11]
```diff
@@ -301,11 +301,11 @@ func (p *Provider) getLaunchTemplateConfigs(ctx context.Context, nodeClass *v1be
 	if err != nil {
 		return nil, fmt.Errorf("getting launch templates, %w", err)
 	}
-	for launchTemplateName, instanceTypes := range launchTemplates {
+	for _, launchTemplate := range launchTemplates {
 		launchTemplateConfig := &ec2.FleetLaunchTemplateConfigRequest{
-			Overrides: p.getOverrides(instanceTypes, zonalSubnets, scheduling.NewNodeSelectorRequirements(nodeClaim.Spec.Requirements...).Get(v1.LabelTopologyZone), capacityType),
+			Overrides: p.getOverrides(instanceTypes, zonalSubnets, scheduling.NewNodeSelectorRequirements(nodeClaim.Spec.Requirements...).Get(v1.LabelTopologyZone), capacityType, launchTemplate.ImageID),
```
Comment (jonathan-innis) [inline pkg/providers/instance/instance.go:306]:
```suggestion
			Overrides: p.getOverrides(launchTemplate.InstanceTypes, zonalSubnets, scheduling.NewNodeSelectorRequirements(nodeClaim.Spec.Requirements...).Get(v1.LabelTopologyZone), capacityType, launchTemplate.ImageID),
```

Subtle bug but I tracked it down. You were getting failures here due to the following error

```console
"launching machine, creating instance, creating fleet InvalidFleetConfig: The fleet configuration contains duplicate instance pools.\n\tstatus code: 400"
```

Looking into it, I saw that we were passing the same overrides for each launch template that we were launching with.

Comment (gfcroft) [reply pkg/providers/instance/instance.go:306]: ah thanks a million for finding this! I was starting to have to work out how to setup e2e tests against my own cluster!

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot
